### PR TITLE
Added flag enums support

### PR DESCRIPTION
**Pull request by daler-sz**
#197

### CODE REVIEW

File: src/adaptix/_internal/morphing/load_error.py [hunk 1]
```diff
@@ -130,6 +130,14 @@ class BadVariantError(LoadError):
     input_value: Any
 
 
+@custom_exception
+@dataclass(eq=False)
+class MultipleBadVariantError(LoadError):
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/load_error.py:135]: New exception naming policy suppose that name signaling error shouldn't have ``Error`` suffix.
Old exception will be renamed. It's ok that name of new exception is inconsistent with old one.

File: src/adaptix/_internal/morphing/load_error.py [hunk 2]
```diff
@@ -130,6 +130,14 @@ class BadVariantError(LoadError):
...
+    allowed_values: Iterable[Any]
+    input_values: Iterable[Any]
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/load_error.py:137]: All ``LoadError`` must have ``input_value: Any`` field (not ``input_values``). Maybe for this case it can have ``Iterable[Any]`` type

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 3]
```diff
@@ -155,3 +157,85 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
 
     def _provide_dumper(self, mediator: Mediator, request: DumperRequest) -> Dumper:
         return _enum_exact_value_dumper
+
+
+def _extract_common_cases_from_flag(enum: Type[Flag]) -> Iterable[Flag]:
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/enum_provider.py:162]: In the rest of the code such flag cases name ``compound``

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 4]
```diff
@@ -155,3 +157,85 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
...
+    cases = enum.__members__.values()
+    result = []
+    for case in cases:
+        if not math.log2(case.value) % 1:
+            result.append(case)
+    return result
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/enum_provider.py:168]: I think that list comprehension will be look nice here

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 5]
```diff
@@ -155,3 +157,85 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
...
+                allowed_values=variants,
+                input_values=process_data,
+                invalid_values=bad_variants
+            )
+
+        return result
+
+    def _provide_loader(self, mediator: Mediator, request: LoaderRequest) -> Loader:
+        enum = get_type_from_request(request)
+        return partial(self._flag_loader, enum=enum)
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/enum_provider.py:230]: Returning partial will hold reference to provider, this prevents from producing standalone file with loader (file that knows nothing about retorts, this is planned feature).
Also permanent lookup through self attributes slow down code.

Loader should be closure capturing variables storing self atrributes

**Review by zhPavel — CHANGES_REQUESTED** [review]

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 6]
```diff
@@ -155,3 +157,85 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
...
+
+        self._loader = _enum_exact_value_loader if by_exact_value else _enum_name_loader
+        self._dumper = _enum_exact_value_dumper if by_exact_value else _enum_name_dumper
+
+    def _get_cases(self, enum: Type[Flag]):
+        if self._allow_compound:
+            return enum.__members__.values()
+        return _extract_common_cases_from_flag(enum)
+
+    def _flag_loader(self, data: Union[int, Iterable[Union[int, str]]], enum: Type[Flag]) -> Flag:
```
Comment (daler-sz) [inline src/adaptix/_internal/morphing/enum_provider.py:192]: this typehint is wrong. should be `Union[int, Iterable[int], Iterable[str]]`

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 7]
```diff
@@ -155,3 +210,97 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
...
+            process_data = [data]
+        else:
+            process_data = list(data)
+
+        if not self._allow_duplicates:
+            if len(process_data) != len(set(process_data)):
+                raise ValueLoadError(
+                    msg=f"Duplicates in {enum} loader are not allowed",
+                    input_value=process_data
+                )
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/enum_provider.py:258]: Generated loader should not hold references to provider (should not use ``self``)

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 8]
```diff
@@ -155,3 +210,97 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
...
+            if not self._allow_single_value:
+                raise TypeLoadError(expected_type=Iterable[str], input_value=data)
+            process_data = [data]
+        else:
+            process_data = list(data)
+
+        if not self._allow_duplicates:
+            if len(process_data) != len(set(process_data)):
+                raise ValueLoadError(
+                    msg=f"Duplicates in {enum} loader are not allowed",
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/enum_provider.py:256]: API user knows nothing about what loader and enum class is it. Also I think it should be separate error class like 
```python

@custom_exception
@dataclass(eq=False)
class DuplicatedValues:
    input_value: Any
```

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 9]
```diff
@@ -155,3 +210,97 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
...
+        if not issubclass(enum, Flag):
+            raise CannotProvide
+
+        def flag_dumper(value: Flag) -> Iterable[Hashable]:
+            cases = self._get_cases(enum)
+            mapping = self._mapping_generator.generate_for_dumping(cases)
+
+            if value in cases:
+                return [mapping[value]]
+            return [mapping[case] for case in cases if case in value]
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/enum_provider.py:304]: Is special case for non-compound value worth it?

**Review by zhPavel — CHANGES_REQUESTED** [review]

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 10]
```diff
@@ -155,3 +210,97 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
...
+        return process_data
+
+    def _provide_loader(self, mediator: Mediator, request: LoaderRequest) -> Loader:
+        enum = get_type_from_request(request)
+
+        if not issubclass(enum, Flag):
+            raise CannotProvide
+
+        def _flag_loader(data: Union[int, Iterable[int], Iterable[str]]) -> Flag:
+            process_data = self._get_loader_process_data(data, enum)
```
[the remark this answers is not in the document]
Comment (daler-sz) [reply src/adaptix/_internal/morphing/enum_provider.py:269]: The problem is that the linter complains about cognitive complexity

[the remark this answers is not in the document]
Comment (zhPavel) [reply src/adaptix/_internal/morphing/enum_provider.py:269]: Supress the linter error)

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 11]
```diff
@@ -155,3 +210,97 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
...
+    def _get_loader_process_data(self, data: Iterable[str], enum: Type[Flag]) -> Sequence[str]:
+        ...
+
+    def _get_loader_process_data(self, data, enum):
+        if isinstance(data, (str, int)):
+            if not self._allow_single_value:
+                raise TypeLoadError(expected_type=Iterable[str], input_value=data)
+            process_data = [data]
+        else:
+            process_data = list(data)
```
[the remark this answers is not in the document]
Comment (daler-sz) [reply src/adaptix/_internal/morphing/enum_provider.py:251]: I don't understand this snippet. Firstly, it seems to me that you forgot some `not`'s. Secondly, why we should check type with `type`, shouldn't we interpret all string subtypes as single values?

[the remark this answers is not in the document]
Comment (zhPavel) [reply src/adaptix/_internal/morphing/enum_provider.py:251]: In fact, builtin loader exclude str from Iterable only if strinct_coercion is enabled. But I don’t see the point in considering in this case that str is also an Iterable, regardless of strcit_coercion

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 12]
```diff
@@ -155,3 +215,134 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
...
+
+        cases = self._get_cases(enum)
+        mapping = self._mapping_generator.generate_for_loading(cases)
+        variants = list(mapping.keys())
+        zero_case = enum(0)
+
+        def flag_loader(data) -> Flag:
+            if isinstance(data, Iterable) and type(data) is not str:  # pylint: disable=unidiomatic-typecheck
+                process_data = tuple(data)
+            else:
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/enum_provider.py:284]: There is common policy prevent interpreting dict as iterable of strings if strict_coercion is True

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 13]
```diff
@@ -155,3 +206,131 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
...
+            )
+
+        def flag_loader(data):
+            if type(data) is not int:   # pylint: disable=unidiomatic-typecheck
+                raise TypeLoadError(int, data)
+
+            if not 0 <= data <= flag_mask:
+                raise OutOfRange(0, flag_mask, data)
+
+            return enum(data)
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/enum_provider.py:239]: A comment is needed here that would explain that the data has already been validated and the enum cannot raise an error

File: src/adaptix/_internal/morphing/enum_provider.py [hunk 14]
```diff
@@ -155,3 +206,131 @@ def _get_exact_value_to_member(self, enum: Type[Enum]) -> Optional[Mapping[Any,
...
+                is_terminal=True,
+                is_demonstrative=True,
+            )
+
+        def flag_loader(data):
+            if type(data) is not int:   # pylint: disable=unidiomatic-typecheck
+                raise TypeLoadError(int, data)
+
+            if not 0 <= data <= flag_mask:
+                raise OutOfRange(0, flag_mask, data)
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/enum_provider.py:237]: Operators precedence are not obvious at this case. Something like ``data < 0 or data > flag_mask`` would be more straightforward

File: src/adaptix/_internal/morphing/facade/retort.py [hunk 15]
```diff
@@ -79,6 +79,7 @@ class FilledRetort(OperatingRetort, ABC):
         IsoFormatProvider(time),
         SecondsTimedeltaProvider(),
 
+        FlagByExactValueProvider(),
         EnumExactValueProvider(),  # it has higher priority than scalar types for Enum with mixins
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/facade/retort.py:83]: I think that using high-level factories here shouble become a standard (``flag_by_exact_value`` and ``enum_by_exact_value``)

File: tests/unit/morphing/test_enum_provider.py [hunk 16]
```diff
@@ -152,3 +179,190 @@ def test_value_provider(strict_coercion, debug_trail):
...
+    dumper = retort.get_dumper(FlagEnum)
+    assert dumper(FlagEnum.CASE_ONE) == 1
+    assert dumper(FlagEnum.CASE_ONE | FlagEnum.CASE_FOUR) == 5
+
+    raises_exc(
+        OutOfRange(0, 15, 16), lambda: loader(16)
+    )
+    raises_exc(
+        OutOfRange(0, 15, -1), lambda: loader(-1)
+    )
```
Comment (zhPavel) [inline tests/unit/morphing/test_enum_provider.py:204]: Need tests with data of bad type (string, list, dict, None)

File: tests/unit/morphing/test_enum_provider.py [hunk 17]
```diff
@@ -152,3 +179,190 @@ def test_value_provider(strict_coercion, debug_trail):
...
+                allow_duplicates=allow_duplicates,
+                allow_compound=allow_compound
+            )
+        ]
+    )
+
+    loader = retort.get_loader(FlagEnum)
+    assert loader(["CASE_ONE"]) == FlagEnum.CASE_ONE
+    assert loader(["CASE_ONE", "CASE_TWO"]) == FlagEnum.CASE_THREE
+    assert loader(["CASE_ONE", "CASE_FOUR"]) == FlagEnum.CASE_ONE | FlagEnum.CASE_FOUR
```
Comment (zhPavel) [inline tests/unit/morphing/test_enum_provider.py:266]: Need tests with data of bad type (string, int, dict, None)

**Review by zhPavel — CHANGES_REQUESTED** [review]

**Comment by zhPavel** [comment]
The feature looks almost complete. But there is need a documentation at specific-types-behavior.rst file and docstring of high-level factory ``flag_by_list_using_name`` should contain desctiprion of limitations.

It is also necessary to add a fragment of the changelog with a description of the new feature.
You can find instruction to creating such fragments in docs/changelog/fragments/README.rst

**Review by zhPavel — CHANGES_REQUESTED** [review]

**Review by zhPavel — APPROVED** [review]
